devlib/instrument/schedstats.py

```diff
@@ -0,0 +1,224 @@
+                self.old_sysctl_value = self.target.read_int(self.sysctl_path)
+                self.target.write_value(self.sysctl_path, 1)
+            else:
+                try:
+                    cmdline = self.target.read_value('/proc/cmdline')
+                except TargetError:
+                    raise TargetError(
+                        "Couldn't verify that schedstats is enabled. "
+                        "Enabling CONFIG_PROC_SYSCTL will probably help")
+                if "schedstats=enable" not in cmdline:
+                    raise TargetError(
+                        "schedstats is compiled into the kernel but not enabled at runtime. "
+                        "Enable CONFIG_PROC_SYSCTL or add schedstats=enable to the cmdline.")
+
+        # Take a sample of the schedstat file to figure out which channels to
+        # create.
+        # We'll create a site for each CPU and a site for each sched_domain.
+        for site, measures in self._get_sample().iteritems():
+            if site.startswith('cpu'):
+                measurement_category = 'schedstat_cpu'
+            else:
+                measurement_category = 'schedstat_domain'
+
+            for measurement_name in measures.keys():
+                measurement_type = MeasurementType(
+                    measurement_name, '', measurement_category)
+                self.add_channel(site=site,
+                                 measure=measurement_type)
+
+    def teardown(self):
+        if self.old_sysctl_value is not None:
+            self.target.write_value(self.sysctl_path, self.old_sysctl_value)
+
+    def _get_sample(self):
+        lines = self.target.read_value(self.schedstat_path).splitlines()
+        ret = OrderedDict()
+
+        # Example /proc/schedstat contents:
+        #
+        # version 15
+        # timestamp <timestamp>
+        # cpu0 <cpu fields>
+        # domain0 <domain fields>
+        # domain1 <domain fields>
+        # cpu1 <cpu_fields>
+        # domain0 <domain fields>
+        # domain1 <domain fields>
+
+        curr_cpu = None
+        for line in lines[2:]:
+            tokens = line.split()
+            if tokens[0].startswith('cpu'):
+                curr_cpu = tokens[0]
+                site = curr_cpu
+                measures = CPU_MEASURES
+                tokens = tokens[1:]
+            elif tokens[0].startswith('domain'):
+                if not curr_cpu:
+                    raise TargetError(
+                        'Failed to parse schedstats, found domain before CPU')
+                # We'll name the site for the domain like "cpu0domain0"
+                site = curr_cpu + tokens[0]
+                measures = DOMAIN_MEASURES
+                tokens = tokens[2:]
+            elif tokens[0] == 'eas':
+                # This line is added by EAS features. We don't yet parse it as
+                # it might not be stable.
+                continue
+            else:
+                self.logger.warning(
```

Comment:
Dunno if it really can make sense... but... can we think about generating at run-time a category which has the name of this line's first column and a "generic list of measures", e.g. measure1, measure2, measure3... etc..

This will allows to parse Android specific fileds and then map the measures to proper names in the client code.

Reply:
Yeah that seems like a sensible idea (although in fact it might be easier to just add the EAS-specific schedstats?). Do you mind if we leave that for a separate feature though?